### PR TITLE
Step metrics

### PR DESCRIPTION
**Pull request by catscanner**

### CODE REVIEW

File: slingshot/core/reporter.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    :param fn: the function to be decorated
+    :type fn: C{func}
+
+    :return: the decorated function
+    :rtype: C{func}
+    """
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        counter("internal").inc()
+        if gauge(metric_name).get_value() < counter("internal").get_count() or isNaN(gauge(metric_name).get_value()):
```
Comment (balintzs) [inline slingshot/core/reporter.py]: Please add comments explaining what is happening here.

File: slingshot/core/step.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from pyformance import counter
+from .reporter import MetricReporter, count_calls
```
Comment (balintzs) [inline slingshot/core/step.py]: Do you need to import anything else than `count_calls`?

File: slingshot/core/reporter.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            gauge(metric_name).set_value(counter("internal").get_count())
+        try:
+            output = fn(*args, **kwargs)
+        finally:
+            counter("internal").dec()
+        return output
+    return wrapper
+
+def isNaN(num):
+    return num != num
```
Comment (balintzs) [inline slingshot/core/reporter.py]: This will always return `False`.

Comment (catscanner) [inline slingshot/core/reporter.py]: Interestingly float('nan') != float('nan') returns True, but I found math.isnan which I think is a better thing to use.

File: slingshot/core/reporter.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    :param fn: the function to be decorated
+    :type fn: C{func}
+
+    :return: the decorated function
+    :rtype: C{func}
+    """
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        # increment internal counter whilst the step runs
+        counter("internal").inc()
```
Comment (balintzs) [inline slingshot/core/reporter.py]: What happens to internal counter when this decorator is used for more than one metric?

**Comment by balintzs** [comment]
LGTM

**Pull request merged**